### PR TITLE
[CSL-909] Refactor: move Tx-related utilities out of WALLET_MODE

### PR DESCRIPTION
**Pull request by neongreen**
We used to have `MonadBalances`, `MonadTxHistory`, `sendTx`, etc. only available in WALLET_MODE, even though they didn't depend on wallet and they were useful outside of WALLET_MODE (e.g. for the smartgen). This PR moves them from Pos.Wallet.WalletMode into the following modules (better name suggestions are welcome):

* Pos.Balances         – `MonadBalances`
* Pos.TxHistory        – `MonadTxHistory`
* Pos.Util.Tx          – transaction construction (e.g. `makePubKeyTx`)
* Pos.Communication.Tx – `TxMode` and transaction sending (e.g. `submitTx`)

This lets us
a) build the smartgen without WALLET_MODE
b) compile ~40 modules less on deployment
c) avoid having linker errors (because otherwise wallet-related modules are compiled even without the wallet)

### CODE REVIEW

File: src/Pos/TxHistory.hs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+-- | Check if given 'Address' is one of the receivers of 'Tx'
+hasReceiver :: Tx -> Address -> Bool
+hasReceiver UnsafeTx {..} addr = any ((== addr) . txOutAddress) _txOutputs
+
+-- | Given some 'Utxo', check if given 'Address' is one of the senders of 'Tx'
+hasSender :: MonadUtxoRead m => Tx -> Address -> m Bool
+hasSender UnsafeTx {..} addr = anyM hasCorrespondingOutput $ toList _txInputs
+  where hasCorrespondingOutput txIn =
+            fmap toBool $ fmap ((== addr) . txOutAddress . toaOut) <$> utxoGet txIn
```
Comment (ChShersh) [inline src/Pos/TxHistory.hs]: We have `<<$>>` in `universum` for `fmap . fmap`

File: src/Pos/TxHistory.hs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    lift (MaybeT $ return $ topsortTxs (view _1) txs) >>=
+    foldlM step DL.empty
+  where
+    step ls (WithHash tx txId, _wit, dist) = do
+        let isIncoming = tx `hasReceiver` addr
+        isOutgoing <- tx `hasSender` addr
+        let allToAddr = all ((== addr) . txOutAddress) $ _txOutputs tx
+            isToItself = isOutgoing && allToAddr
+        if isOutgoing || isIncoming
+            then do
```
Comment (ChShersh) [inline src/Pos/TxHistory.hs]: This formatting of `then do` and next block w/o indentation makes me sad :(

File: src/Pos/TxHistory.hs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+            applyTxToUtxo (WithHash tx txId) dist
+            identity %= filterUtxoByAddr addr
+
+            -- Workaround to present A to A transactions as a pair of self-canceling
+            -- transactions in history
+            let resEntry = THEntry txId tx isOutgoing Nothing
+                resList = if isToItself
+                          then DL.fromList [resEntry & thIsOutput .~ False, resEntry]
+                          else DL.singleton resEntry
```
Comment (ChShersh) [inline src/Pos/TxHistory.hs]: Maybe implement `one` for `DList` somewhere and other utilities? I see a lot of usages of `DL` specific functions.

File: src/Pos/TxHistory.hs [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    -> TxSelectorT m [TxHistoryEntry]
+deriveAddrHistoryPartial hist addr chain =
+    DL.toList <$> foldrM updateAll (DL.fromList hist) chain
+  where
+    updateAll (Left _) hst = pure hst
+    updateAll (Right blk) hst = do
+        txs <- getRelatedTxs addr $
+                   map (over _1 withHash) (blk ^. blockTxas)
+        let difficulty = blk ^. difficultyL
+            txs' = map (thDifficulty .~ Just difficulty) txs
```
Comment (ChShersh) [inline src/Pos/TxHistory.hs]: `thDifficulty .~ Just difficulty` is `thDifficulty ?~ difficulty`

File: src/Pos/TxHistory.hs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+class Monad m => MonadTxHistory m where
+    getTxHistory
+        :: SscHelpersClass ssc
+        => Tagged ssc (Address -> Maybe (HeaderHash, Utxo) -> m TxHistoryAnswer)
+    saveTx :: (TxId, TxAux) -> m ()
+
+    default getTxHistory
+        :: (SscHelpersClass ssc, MonadTrans t, MonadTxHistory m', t m' ~ m)
+        => Tagged ssc (Address -> Maybe (HeaderHash, Utxo) -> m TxHistoryAnswer)
+    getTxHistory = fmap (fmap lift) <$> getTxHistory
```
Comment (ChShersh) [inline src/Pos/TxHistory.hs]: Triple `fmap` !!! Maybe use `<<$>>` also here

File: src/Pos/TxHistory.hs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+#endif
+         , PC.WithNodeContext s m) =>
+         MonadTxHistory (TxpHolder m) where
+    getTxHistory :: forall ssc. SscHelpersClass ssc
+                 => Tagged ssc (Address -> Maybe (HeaderHash, Utxo) -> TxpHolder m TxHistoryAnswer)
+    getTxHistory = Tagged $ \addr mInit -> do
+        tip <- GS.getTip
+
+        let getGenUtxo = filterUtxoByAddr addr . PC.ncGenesisUtxo <$> PC.getNodeContext
+        (bot, genUtxo) <- maybe ((,) <$> GS.getBot <*> getGenUtxo) pure mInit
```
Comment (ChShersh) [inline src/Pos/TxHistory.hs]: You can use [`whenNothingM`](https://github.com/serokell/universum/blob/9edb5d374ec1fb6bb873dacc22a013f1eacd9c67/src/Monad/Maybe.hs#L39) here

File: src/Pos/TxHistory.hs [hunk 7]
```diff
@@ -0,0 +1,241 @@
+{-# LANGUAGE CPP                  #-}
+{-# LANGUAGE InstanceSigs         #-}
+{-# LANGUAGE RankNTypes           #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TemplateHaskell      #-}
```
Comment (ChShersh) [inline src/Pos/TxHistory.hs]: I didn't notice `TemplateHaskell` usages. Maybe it's redundant extension here or maybe I didn't pay enough attention. As well as `InstanceSigs`

File: src/Pos/Balances.hs [hunk 8]
```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE TypeFamilies #-}
+
+module Pos.Balances
```
Comment (ChShersh) [inline src/Pos/Balances.hs]: I think this should be somewhere is `Pos.Txp`, like, `Pos.Txp.MonadBalances`. I don't like this module is not related to any subpart, like `Txp` or `Wallet`

File: src/Pos/TxHistory.hs [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Pos.TxHistory
```
Comment (ChShersh) [inline src/Pos/TxHistory.hs]: `Pos.Txp.TxHistory` is better name, I think. Again, better to put inside some subpart, and `Txp` seems like reasonable part here.

File: src/Pos/Util/Tx.hs [hunk 10]
```diff
@@ -0,0 +1,142 @@
+-- | Pure functions for operations with transactions
+
+module Pos.Util.Tx
```
Comment (ChShersh) [inline src/Pos/Util/Tx.hs]: From my point of `view` module `Util` should contain some general purpose functions which can be used everywhere, like some logging utility functions. All these functions are very specific. I think it's better to put into `Pos.Txp.Toil.Pure` but I don't remember quite well what is `Toil`. Maybe @gromakovsky can add something.

File: src/Pos/TxHistory.hs (hunk 7 above)
Comment (neongreen) [inline src/Pos/TxHistory.hs]: I use `makeLenses` there

Comment (neongreen) [inline src/Pos/TxHistory.hs]: As for `InstanceSigs`, I don't actually remember what it does but stylish-haskell refused to process the file until I added it.

File: src/Pos/TxHistory.hs (hunk 9 above)
Comment (neongreen) [inline src/Pos/TxHistory.hs]: Okay, will rename.

File: src/Pos/TxHistory.hs (hunk 4 above)
Comment (neongreen) [inline src/Pos/TxHistory.hs]: Sure, but I suspect that not everyone is aware of `?~` and it's not even that useful to force its usage

File: src/Pos/TxHistory.hs (hunk 6 above)
Comment (neongreen) [inline src/Pos/TxHistory.hs]: How?

File: src/Pos/Util/Tx.hs (hunk 10 above)
Comment (neongreen) [inline src/Pos/Util/Tx.hs]: I don't know what is `Toil` either, though I think I can remember @gromakovsky explaining it to me :trollface: Okay, waiting for him to comment. (Will refrain from renaming the module until then.)

**Comment by gromakovsky** [comment]
`Pos.TxHistory` and `Pos.MonadBalances` is very bad. `Pos.Txp.TxHistory` and `Pos.Txp.MonadBalances` is better, but I still don't like it.

Personally I want client logic to be separated from node's logic. I think it's a different layer on which node itself shouldn't depend. Currently we are doing some refactorings to add more self-contained parts. I think client logic is a good separate part.

What do you think about the following approaches?
• Put all client logic which we want to have in `cardano-sl` library into `Pos.Client.*` (e. g. `Pos.Client.Txp.MonadBalances`, etc.).
• Put client logic of component X into `Pos.X.Client` (e. g. `Pos.Txp.Client.MonadBalances`).

Regarding `Pos.Util.Tx`: I think it's a wrong approach. We had a conversion with Misha long ago and decided that if there is a component X and activity Y (in our case `X = Txp` and `Y = Util`), then:
• if X is small, we should have `Pos.Y.X`;
• otherwise we should have `Pos.X.Y`.

`Txp` is definitely big component. Hence it should be at least `Pos.Txp.Util` I think. Or better (imo) `Pos.Client.Txp.Util`.

**Comment by neongreen** [comment]
@gromakovsky 

> Txp is definitely big component. Hence it should be at least Pos.Txp.Util I think. Or better (imo) Pos.Client.Txp.Util.

Okay, I will change it.

> Put all client logic which we want to have in cardano-sl library into Pos.Client.* (e. g. Pos.Client.Txp.MonadBalances, etc.).

I like this approach more than the `Pos.Txp.Client` one. Should I go with it now, or would it be better to leave it for some future refactoring?

**Comment by gromakovsky** [comment]
> I like this approach more than the Pos.Txp.Client one. Should I go with it now, or would it be better to leave it for some future refactoring?

Yes, I like `Pos.Client.X` more than `Pos.X.Client` too.
I think moving few modules related to this PR should take about 10 minutes. If I am right, do it now, otherwise it's fine to leave it for another refactoring.
